src/ir/datalog/program.h

```diff
@@ -49,19 +51,9 @@ class Predicate {
   }
   // Equality is also needed to use a Predicate in a flat_hash_set
   bool operator==(const Predicate& otherPredicate) const {
-    if (this->name() != otherPredicate.name()) {
-      return false;
-    }
-    if (this->sign() != otherPredicate.sign()) {
-      return false;
-    }
-    if (this->args().size() != otherPredicate.args().size()) {
-      return false;
-    }
-    for (uint64_t i = 0; i < this->args().size(); i++) {
-      if (this->args().at(i) != otherPredicate.args().at(i)) return false;
-    }
-    return true;
+    return this->name() == otherPredicate.name() &&
+           this->sign() == otherPredicate.sign() &&
+           absl::c_equal(this->args(), otherPredicate.args());
```

Comment:
While I think this is a good way to simplify `operator==`, it doesn't make progress towards an even bigger refactoring that I think would be valuable. That is, the `operator==` and `operator<` have separate implementations, but it would be nice to just define the order on predicates with a `strcmp`-style `compare` function and have `operator==`'s body be just `compare(other) == 0` and `operator<`'s body be just `compare(other) < 0`. Should we just create the `compare` function?

Reply:
Relevant issue: https://github.com/google-research/raksha/issues/364

Reply:
@markww Just curious..Does compare(other) compare between 2 predicates? I can imagine a scenario where we check if 2 predicates are same. What is the use case for operator< ??

Reply:
See here: https://github.com/google-research/raksha/blob/main/src/ir/auth_logic/souffle_emitter.h#L116 . That said, maybe the `Predicate`s should be stored in a `vector` rather than a `set`, which would make the canonical order just "the order in which the `Predicate`s were encountered", which would make `<` unnecessary.

Reply:
I think there is some implicit de-duping based on `name`. In any case that use needs to be fixed and we should get rid of `operator <`. I will add the discussion here to #364 and resolving this comment.

Reply:
Wait, before you do, are we sure that `this->args() == otherPredicate.args()` wouldn't work?

Reply:
Yes, actually: `args` is a `std::vector` which, if its value type has an `operator==`, also has an `operator==`. So we should probably just use `std::vector` `operator==` instead of `c_equal`.